ZFS replication compatibility and per-dataset feature activation

allanjude opened the pull request.

allanjude committed:
ZSTD replication compatibility improvements

If the dataset being sent has the ZSTD feature activated, and
the replication compression flag is specified, we set the ZSTD
replication feature flag, so the receiving side will reject the
stream if it does not support ZSTD.

Only send ZSTD compressed embedded block pointers if the ZSTD
replication flag is already set.

`zfs send -c` will send ZSTD compressed blocks
`zfs send -e` will send embedded LZ4 block pointers but convert
    ZSTD eBP's into regular blocks
`zfs send -ec` will send ZSTD compressed blocks and eBPs

Signed-off-by: Allan Jude <allanjude@freebsd.org>

Crow-Control commented:
I'm not well versed enough in the code side of things to comment on it.
But design wise it solves some of the remaining issues as promissed....

allanjude commented:
Yeah, the replication stuff is as-discussed on the call.

The feature flag thing solves the problem one of you pointed out previously:
zfs set compress=zstd
zpool export
zpool import (on 0.8.x machine)
zfs get compress -> panic/abort

Crow-Control commented:
Indeed, it looks like a clean and simple fix for our most our problems...
Anyway: @c0d3z3r0 LGTM....

c0d3z3r0 commented:
great! will merge on monday I guess, as soon as I have sorted out the build problem.

allanjude committed:
Activate per-dataset feature flags when the property is set

If a compression algorithm requires a feature flag, activate the
feature when the property is set, rather than waiting for the first
block to be born.  This is currently only used by zstd but can be
extended as needed.

Signed-off-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Allan Jude <allanjude@freebsd.org>

allanjude committed:
Add a new test to confirm that the ZSTD feature flag is activated

When the compression property on a dataset is set to zstd, the feature
flag is not changed from 'enabled' to 'active' until the first block
is born.

If the pool is exported after `zfs set` but before a block is written
then imported on a system with an older version of ZFS that does not
understand ZSTD, it will cause the userland utilities to panic.

Signed-off-by: Allan Jude <allanjude@freebsd.org>

allanjude committed:
[Squashme] Fix copyright message on new tests

Signed-off-by: Allan Jude <allanjude@freebsd.org>
